File: code_LS.R
## least squares
set.seed(1)
n = 10^4
p = 10^3
X = matrix(rnorm(n * p), n, p)
beta = rep(1, p)
Y = X %*% beta + rnorm(n)

# the worst
system.time(
    solve(t(X) %*% X) %*% t(X) %*% Y
)

# can be better
system.time(
    solve(t(X) %*% X) %*% (t(X) %*% Y)
)

# should be this way
system.time(
    solve(t(X) %*% X, t(X) %*% Y)
)

system.time(lm.fit(X, Y))

## weighted least squares
set.seed(1)
n = 10^4
p = 10^2
X = matrix(rnorm(n * p), n, p)
beta = rep(1, p)
Y = X %*% beta + rnorm(n)

w = runif(n)
W = diag(w)

# never do this
system.time(solve(t(X) %*% W %*% X, t(X) %*% W %*% Y))

# do it this way
system.time({
    wX = w * X
    solve(t(X) %*% wX, t(wX) %*% Y)
})

## leverage scores

# never do this
system.time({
    L1 = diag(X %*% solve(t(X) %*% X, t(X)))
})

# try the following two
library(expm)
system.time({
    XXrt <- sqrtm(t(X) %*% X)
    L2 = colSums((solve(XXrt, t(X))^2))
})

system.time({L3 <- rowSums(svd(X)$u^2)})

File: code_LS.jl
# least squares

using Random
Random.seed!(1)
n, p = 10^4, 10^3
X = randn(n, p);
β = ones(p);
Y = X * β .+ randn(n);

# the first time running is slow
@time β̂ = X'X \ X'Y;
@time β̂ = X'X \ X'Y;

# weighted least squares

n, p = 10^4, 10^2
X = randn(n, p);
β = ones(p);
Y = X * β .+ randn(n);
w = rand(n)

@time begin
    wX = X .* w
    β̂w = wX'X \ wX'Y
end;

function wls(X, Y, w)
    wX = X .* w
    β̂w = wX'X \ wX'Y
end

# pre allocate wX to avoid memory allocation
wX = similar(X);
function wls!(X, Y, w, wX)
    wX .= X .* w
    β̂w = wX'X \ wX'Y
end

rpt = 10
@time for i in 1:rpt β̂w = wls(X, Y, w) end
@time for i in 1:rpt β̂w = wls!(X, Y, w, wX) end

# leverage scores
using LinearAlgebra
# @time L1 = diag(X / (X'X) * X');
@time L2 = sum(abs2, X / √(X'X), dims=2);
@time L3 = sum(abs2, svd(X).U, dims=2);
